src/utilities/include/hashtable.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2022 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+#include <stdint.h>
+
+#include <atomic>
+
+#include "public.h"
+#include "utils.h"
+#include "vec_dh.h"
+
+namespace {
+typedef unsigned long long int Uint64;
+constexpr Uint64 kOpen = std::numeric_limits<Uint64>::max();
+
+template <typename T>
+__host__ __device__ T AtomicCAS(T& target, T compare, T val) {
+#ifdef __CUDA_ARCH__
+  return atomicCAS(&target, compare, val);
+#else
+  std::atomic<T>& tar = reinterpret_cast<std::atomic<T>&>(target);
+  tar.compare_exchange_strong(compare, val);
+  return compare;
+#endif
+}
+
+// https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key
+__host__ __device__ inline Uint64 hash64bit(Uint64 x) {
+  x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9ull;
+  x = (x ^ (x >> 27)) * 0x94d049bb133111ebull;
+  x = x ^ (x >> 31);
+  return x;
+}
+}  // namespace
+
+namespace manifold {
+/** @addtogroup Private
+ *  @{
+ */
+
+template <typename V>
+class HashTableD {
+ public:
+  HashTableD(VecDH<Uint64>& keys, VecDH<V>& values, VecDH<uint32_t>& used,
+             uint32_t step = 1)
+      : step_{step}, keys_{keys}, values_{values}, used_{used} {}
+
+  __host__ __device__ int Size() const { return keys_.size(); }
+
+  __host__ __device__ bool Full() const { return used_[0] * 2 > Size(); }
+
+  __host__ __device__ void Insert(Uint64 key, const V& val) {
+    uint32_t idx = hash64bit(key) & (Size() - 1);
```

Comment:
Since I can't see the diff, is there anything else you changed in this hash table? I can see you added this hash function, which may be good for your new uses, but I think will not be for its original SDF purpose. 

The reason I used a step of 127 is the same reason I used a trivial hash function: this is designed to be a locality-sensitive hash function, and therefor fast for a GPU cache. The threads in a warp are sequential Morton codes, so they are nearby in space, and when they make hash table entries (near the surface) they also hash to nearby (often sequential) locations. This keeps the cache in good shape, but if I collide, then stepping by 1 is nearly always another collision. Better to step far enough to get off the local surface, since we'll probably need a new cache line anyway (and the other threads can probably reuse it). 

I haven't done a proper performance comparison, but I imagine it'd be noticeable for a large SDF running on CUDA.

Reply:
I separated keys and values into two vectors, because for the case of storing u64 key and u32 value, using a single vector with `pair<u64, u32>` will waste 32 bit of memory per element due to padding requirement. For the hash function for SDF, I just run the tests and don't see any performance regression, I am not sure how it will perform for larger SDF instances. I guess maybe I can try to come up with some larger instances and check?

Reply:
I wanted to use template to allow users to choose the hash function, but I somehow cannot get the compiler to accept it.

Reply:
If you try GyroidModule at a higher resolution that should run it long enough for a stress test pretty easily.

Reply:
OK, it seems that the hash function has quite significant impact on the performance. I tested the Gyroid test `edgeLength=0.05`. The run using *random hash* took 1900ms while the run using the identity function took ~1500ms. I tried using different step size but it doesn't seem to matter in this case. I now added the generic hash function in the template parameter (it seems that the compiler is happy when we use c++17, which is already the case).

Reply:
I guess the problem is that we cannot avoid encountering primary clustering especially when we use locality sensitive hashing. Step size of 1, 127, or anything will not give us better running time because we can fall into another cluster after stepping through that many entries.